controllers/machine_controller.go

```diff
@@ -382,18 +382,22 @@ func (r *MachineReconciler) isDeleteNodeAllowed(ctx context.Context, cluster *cl
 		return err
 	}
 
-	// Whether or not it is okay to delete the NodeRef depends on the
+	// Under a kubeadm control plane, whether or not it is okay to delete the NodeRef depends on the
 	// number of remaining control plane members and whether or not this
 	// machine is one of them.
-	switch numControlPlaneMachines := len(util.GetControlPlaneMachines(machines)); {
-	case numControlPlaneMachines == 0:
-		// Do not delete the NodeRef if there are no remaining members of
-		// the control plane.
-		return errNoControlPlaneNodes
-	default:
-		// Otherwise it is okay to delete the NodeRef.
-		return nil
+	if cluster.Spec.ControlPlaneRef.Kind == "KubeadmControlPlane" {
```

Comment:
We can't assume the `Kind` here is always going to be of type `KubeadmControlPlane`, there might be other infrastructure providers, or single-machine control plane nodes (outside of KCP) that are using this feature. The change here makes it opt in only for KCP clusters, which is a behavioral breaking change.

In alternative, we could have a status field that reports the number of available control plane machines, and the machine controller can use it here. We'd have to make this field fail gracefully, to support the current logic in case the field isn't available or it doesn't exist.

If possible, we should try to keep design conversations on the issue itself, wdyt?

Reply:
Happy to have the conversation in the issue if you prefer. I had a thought that a couple lines of code would help start the conversation since I was struggling to understand how this section is being used.